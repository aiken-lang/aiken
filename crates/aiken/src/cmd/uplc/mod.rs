mod decode;
mod encode;
mod eval;
mod fmt;
mod shrink;

use clap::{Subcommand, ValueEnum};

#[derive(Copy, Clone, ValueEnum)]
pub(super) enum Format {
    Name,
    NamedDebruijn,
    Debruijn,
}

/// Commands for working with untyped Plutus-core
#[derive(Subcommand)]
pub enum Cmd {
    Fmt(fmt::Args),
    Eval(eval::Args),
    #[clap(alias = "flat")]
    Encode(encode::Args),
    #[clap(alias = "unflat")]
    Decode(decode::Args),
    #[clap(alias = "optimize")]
    Shrink(shrink::Args),
}

pub fn exec(cmd: Cmd) -> miette::Result<()> {
    match cmd {
        Cmd::Fmt(args) => fmt::exec(args),
        Cmd::Eval(args) => eval::exec(args),
        Cmd::Encode(args) => encode::exec(args),
        Cmd::Decode(args) => decode::exec(args),
        Cmd::Shrink(args) => shrink::exec(args),
    }
}
