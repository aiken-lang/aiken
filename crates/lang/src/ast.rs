use std::{collections::HashMap, fmt, ops::Range, sync::Arc};

use internment::Intern;

use crate::{
    expr::{TypedExpr, UntypedExpr},
    tipo::{self, PatternConstructor, Type, ValueConstructor},
};

pub const CAPTURE_VARIABLE: &str = "_capture";

pub type TypedModule = Module<tipo::Module, TypedDefinition>;
pub type UntypedModule = Module<(), UntypedDefinition>;

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub enum ModuleKind {
    Lib,
    Script,
}

impl ModuleKind {
    pub fn is_script(&self) -> bool {
        matches!(self, ModuleKind::Script)
    }

    pub fn is_lib(&self) -> bool {
        matches!(self, ModuleKind::Lib)
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Module<Info, Definitions> {
    pub name: String,
    pub docs: Vec<String>,
    pub type_info: Info,
    pub definitions: Vec<Definitions>,
    pub kind: ModuleKind,
}

impl UntypedModule {
    pub fn dependencies(&self) -> Vec<(String, Span)> {
        self.definitions()
            .flat_map(|def| {
                if let Definition::Use {
                    location, module, ..
                } = def
                {
                    Some((module.join("/"), *location))
                } else {
                    None
                }
            })
            .collect()
    }

    pub fn definitions(&self) -> impl Iterator<Item = &UntypedDefinition> {
        self.definitions.iter()
    }

    pub fn into_definitions(self) -> impl Iterator<Item = UntypedDefinition> {
        self.definitions.into_iter()
    }
}

pub type TypedDefinition = Definition<Arc<Type>, TypedExpr, String, String>;
pub type UntypedDefinition = Definition<(), UntypedExpr, (), ()>;

#[derive(Debug, Clone, PartialEq)]
pub enum Definition<T, Expr, ConstantRecordTag, PackageName> {
    Fn {
        arguments: Vec<Arg<T>>,
        body: Expr,
        doc: Option<String>,
        location: Span,
        name: String,
        public: bool,
        return_annotation: Option<Annotation>,
        return_type: T,
    },

    TypeAlias {
        alias: String,
        annotation: Annotation,
        doc: Option<String>,
        location: Span,
        parameters: Vec<String>,
        public: bool,
        tipo: T,
    },

    DataType {
        constructors: Vec<RecordConstructor<T>>,
        doc: Option<String>,
        location: Span,
        name: String,
        opaque: bool,
        parameters: Vec<String>,
        public: bool,
        typed_parameters: Vec<T>,
    },

    Use {
        as_name: Option<String>,
        location: Span,
        module: Vec<String>,
        package: PackageName,
        unqualified: Vec<UnqualifiedImport>,
    },

    ModuleConstant {
        doc: Option<String>,
        location: Span,
        public: bool,
        name: String,
        annotation: Option<Annotation>,
        value: Box<Constant<T, ConstantRecordTag>>,
        tipo: T,
    },
}

pub type TypedConstant = Constant<Arc<Type>, String>;
pub type UntypedConstant = Constant<(), ()>;

#[derive(Debug, Clone, PartialEq)]
pub enum Constant<T, RecordTag> {
    Int {
        location: Span,
        value: String,
    },

    String {
        location: Span,
        value: String,
    },

    Pair {
        location: Span,
        elements: Vec<Self>,
    },

    List {
        location: Span,
        elements: Vec<Self>,
        tipo: T,
    },

    Record {
        location: Span,
        module: Option<String>,
        name: String,
        args: Vec<CallArg<Self>>,
        tag: RecordTag,
        tipo: T,
        field_map: Option<FieldMap>,
    },

    ByteString {
        location: Span,
        // segments: Vec<BitStringSegment<Self, T>>,
    },

    Var {
        location: Span,
        module: Option<String>,
        name: String,
        constructor: Option<Box<ValueConstructor>>,
        tipo: T,
    },
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CallArg<A> {
    pub label: Option<String>,
    pub location: Span,
    pub value: A,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FieldMap {
    pub arity: usize,
    pub fields: HashMap<String, usize>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct RecordConstructor<T> {
    pub location: Span,
    pub name: String,
    pub arguments: Vec<RecordConstructorArg<T>>,
    pub documentation: Option<String>,
    pub sugar: bool,
}

#[derive(Debug, Clone, PartialEq)]
pub struct RecordConstructorArg<T> {
    pub label: Option<String>,
    // ast
    pub annotation: Annotation,
    pub location: Span,
    pub tipo: T,
    pub doc: Option<String>,
}

pub type UntypedArg = Arg<()>;

#[derive(Debug, Clone, PartialEq)]
pub struct Arg<T> {
    pub arg_name: ArgName,
    pub location: Span,
    pub annotation: Option<Annotation>,
    pub tipo: T,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ArgName {
    Discard {
        name: String,
        location: Span,
    },
    LabeledDiscard {
        label: String,
        name: String,
        location: Span,
    },
    Named {
        name: String,
        location: Span,
    },
    NamedLabeled {
        name: String,
        label: String,
        location: Span,
    },
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct UnqualifiedImport {
    pub location: Span,
    pub name: String,
    pub as_name: Option<String>,
    pub layer: Layer,
}

// TypeAst
#[derive(Debug, Clone, PartialEq)]
pub enum Annotation {
    Constructor {
        location: Span,
        module: Option<String>,
        name: String,
        arguments: Vec<Self>,
    },

    Fn {
        location: Span,
        arguments: Vec<Self>,
        ret: Box<Self>,
    },

    Var {
        location: Span,
        name: String,
    },

    Tuple {
        location: Span,
        elems: Vec<Self>,
    },

    Hole {
        location: Span,
        name: String,
    },
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Layer {
    Value,
    Type,
}

impl Default for Layer {
    fn default() -> Self {
        Layer::Value
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum BinOp {
    // Boolean logic
    And,
    Or,

    // Equality
    Eq,
    NotEq,

    // Order comparison
    LtInt,
    LtEqInt,
    GtEqInt,
    GtInt,

    // Maths
    AddInt,
    SubInt,
    MultInt,
    DivInt,
    ModInt,
}

pub type UntypedPattern = Pattern<(), ()>;
pub type TypedPattern = Pattern<PatternConstructor, Arc<Type>>;

#[derive(Debug, Clone, PartialEq)]
pub enum Pattern<Constructor, Type> {
    Int {
        location: Span,
        value: String,
    },

    Float {
        location: Span,
        value: String,
    },

    String {
        location: Span,
        value: String,
    },

    /// The creation of a variable.
    /// e.g. `assert [this_is_a_var, .._] = x`
    Var {
        location: Span,
        name: String,
    },

    /// A reference to a variable in a bit string. This is always a variable
    /// being used rather than a new variable being assigned.
    VarUsage {
        location: Span,
        name: String,
        tipo: Type,
    },

    /// A name given to a sub-pattern using the `as` keyword.
    /// e.g. `assert #(1, [_, _] as the_list) = x`
    Assign {
        name: String,
        location: Span,
        pattern: Box<Self>,
    },

    /// A pattern that binds to any value but does not assign a variable.
    /// Always starts with an underscore.
    Discard {
        name: String,
        location: Span,
    },

    List {
        location: Span,
        elements: Vec<Self>,
        tail: Option<Box<Self>>,
    },

    /// The constructor for a custom type. Starts with an uppercase letter.
    Constructor {
        location: Span,
        name: String,
        arguments: Vec<CallArg<Self>>,
        module: Option<String>,
        constructor: Constructor,
        with_spread: bool,
        tipo: Type,
    },

    Tuple {
        location: Span,
        elems: Vec<Self>,
    },
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum AssignmentKind {
    Let,
    Assert,
}

pub type MultiPattern<PatternConstructor, Type> = Vec<Pattern<PatternConstructor, Type>>;

pub type UntypedMultiPattern = MultiPattern<(), ()>;
pub type TypedMultiPattern = MultiPattern<PatternConstructor, Arc<Type>>;

pub type TypedClause = Clause<TypedExpr, PatternConstructor, Arc<Type>, String>;

pub type UntypedClause = Clause<UntypedExpr, (), (), ()>;

#[derive(Debug, Clone, PartialEq)]
pub struct Clause<Expr, PatternConstructor, Type, RecordTag> {
    pub location: Span,
    pub pattern: MultiPattern<PatternConstructor, Type>,
    pub alternative_patterns: Vec<MultiPattern<PatternConstructor, Type>>,
    pub guard: Option<ClauseGuard<Type, RecordTag>>,
    pub then: Expr,
}

#[derive(Debug, Clone, PartialEq)]
pub enum ClauseGuard<Type, RecordTag> {
    Equals {
        location: Span,
        left: Box<Self>,
        right: Box<Self>,
    },

    NotEquals {
        location: Span,
        left: Box<Self>,
        right: Box<Self>,
    },

    GtInt {
        location: Span,
        left: Box<Self>,
        right: Box<Self>,
    },

    GtEqInt {
        location: Span,
        left: Box<Self>,
        right: Box<Self>,
    },

    LtInt {
        location: Span,
        left: Box<Self>,
        right: Box<Self>,
    },

    LtEqInt {
        location: Span,
        left: Box<Self>,
        right: Box<Self>,
    },

    Or {
        location: Span,
        left: Box<Self>,
        right: Box<Self>,
    },

    And {
        location: Span,
        left: Box<Self>,
        right: Box<Self>,
    },

    Var {
        location: Span,
        tipo: Type,
        name: String,
    },

    TupleIndex {
        location: Span,
        index: u64,
        tipo: Type,
        tuple: Box<Self>,
    },

    Constant(Constant<Type, RecordTag>),
}

pub type TypedIfBranch = IfBranch<TypedExpr>;
pub type UntypedIfBranch = IfBranch<UntypedExpr>;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct IfBranch<Expr> {
    pub condition: Expr,
    pub body: Expr,
    pub location: Span,
}

#[derive(Debug)]
pub struct TypedRecordUpdateArg {
    pub label: String,
    pub location: Span,
    pub value: TypedExpr,
    pub index: usize,
}

#[derive(Debug, Clone, PartialEq)]
pub struct UntypedRecordUpdateArg {
    pub label: String,
    pub location: Span,
    pub value: UntypedExpr,
}

#[derive(Debug, Clone, PartialEq)]
pub struct RecordUpdateSpread {
    pub base: Box<UntypedExpr>,
    pub location: Span,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TodoKind {
    Keyword,
    EmptyFunction,
}

#[derive(Copy, Clone, PartialEq, Eq, Hash, Debug)]
pub struct SrcId(Intern<Vec<String>>);

impl SrcId {
    pub fn empty() -> Self {
        SrcId(Intern::new(Vec::new()))
    }
}

#[derive(Copy, Clone, PartialEq, Eq)]
pub struct Span {
    pub src: SrcId,
    pub start: usize,
    pub end: usize,
}

impl From<Span> for miette::SourceSpan {
    fn from(span: Span) -> Self {
        Self::new(span.start.into(), (span.end - span.start).into())
    }
}

impl Span {
    pub fn empty() -> Self {
        use chumsky::Span;

        Self::new(SrcId::empty(), 0..0)
    }

    pub fn src(&self) -> SrcId {
        self.src
    }

    pub fn range(&self) -> Range<usize> {
        use chumsky::Span;

        self.start()..self.end()
    }

    pub fn union(self, other: Self) -> Self {
        use chumsky::Span;

        assert_eq!(
            self.src, other.src,
            "attempted to union spans with different sources"
        );

        Self {
            start: self.start().min(other.start()),
            end: self.end().max(other.end()),
            ..self
        }
    }
}

impl fmt::Debug for Span {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{:?}:{:?}", self.src, self.range())
    }
}

impl chumsky::Span for Span {
    type Context = SrcId;

    type Offset = usize;

    fn new(context: Self::Context, range: Range<Self::Offset>) -> Self {
        assert!(range.start <= range.end);

        Self {
            src: context,
            start: range.start,
            end: range.end,
        }
    }

    fn context(&self) -> Self::Context {
        self.src
    }

    fn start(&self) -> Self::Offset {
        self.start
    }

    fn end(&self) -> Self::Offset {
        self.end
    }
}
