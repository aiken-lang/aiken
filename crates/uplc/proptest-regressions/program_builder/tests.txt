# Seeds for failure cases proptest has generated in the past. It is
# automatically read and these particular cases re-run before any
# novel cases are generated.
#
# It is recommended to check this file in to source control so that
# everyone who runs the test benefits from these saved cases.
cc db95a6ed96ee7cee987ba4b71bc6ffa9cf5abe295fa57a0eade27ed47335c1a3 # shrinks to maj = 0, min = 0, patch = 0
