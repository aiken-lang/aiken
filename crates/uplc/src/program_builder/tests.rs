use super::*;
use crate::parser;

#[test]
fn build_named__with_const() {
    let code = r"(program
                       11.22.33
                       (con integer 11)
                     )";
    let expected = parser::program(code).unwrap();
    let actual = Builder::new(11, 22, 33).with_constant_int(11).build_named();
    assert_eq!(expected, actual);
}

#[test]
fn build_named__with_different_const() {
    let code = r"(program
                       11.22.33
                       (con integer 22)
                     )";
    let expected = parser::program(code).unwrap();
    let actual = Builder::new(11, 22, 33).with_constant_int(22).build_named();
    assert_eq!(expected, actual);
}

#[test]
fn build_named__with_const_different_version() {
    let code = r"(program
                       44.55.66
                       (con integer 11)
                     )";
    let expected = parser::program(code).unwrap();
    let actual = Builder::new(44, 55, 66).with_constant_int(11).build_named();
    assert_eq!(expected, actual);
}

#[test]
fn build_named__with_lam() {
    let code = r"(program
                       1.2.3
                       (lam i_0 (con integer 1))
                     )";
    let expected = parser::program(code).unwrap();
    let actual = Builder::new(1, 2, 3)
        .with_lambda("i_0")
        .with_constant_int(1)
        .build_named();
    assert_eq!(expected, actual);
}

#[test]
fn build_named__with_nested_lam() {
    let code = r"(program
                       1.2.3
                       (lam i_0 (lam i_1 (con integer 1)))
                     )";
    let expected = parser::program(code).unwrap();
    let actual = Builder::new(1, 2, 3)
        .with_lambda("i_0")
        .with_lambda("i_1")
        .with_constant_int(1)
        .build_named();
    assert_eq!(expected, actual);
}
