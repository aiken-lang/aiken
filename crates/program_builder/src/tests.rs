use crate::Builder;
use uplc::parser;

#[test]
fn build_named__with_const() {
    let code = r"(program
                       11.22.33
                       (con integer 11)
                     )";
    let expected = parser::program(code).unwrap();
    let actual = Builder::default()
        .with_version(11, 22, 33)
        .with_constant_int(11)
        .build_named();
    assert_eq!(expected, actual);
}

#[test]
fn build_named__with_different_const() {
    let code = r"(program
                       11.22.33
                       (con integer 22)
                     )";
    let expected = parser::program(code).unwrap();
    let actual = Builder::default()
        .with_version(11, 22, 33)
        .with_constant_int(22)
        .build_named();
    assert_eq!(expected, actual);
}

#[test]
fn build_named__with_const_different_version() {
    let code = r"(program
                       44.55.66
                       (con integer 11)
                     )";
    let expected = parser::program(code).unwrap();
    let actual = Builder::default()
        .with_version(44, 55, 66)
        .with_constant_int(11)
        .build_named();
    assert_eq!(expected, actual);
}

#[test]
fn build_named__with_lam() {
    let code = r"(program
                       1.2.3
                       (lam i_0 (con integer 1))
                     )";
    let expected = parser::program(code).unwrap();
    let actual = Builder::default()
        .with_version(1, 2, 3)
        .with_lambda(1)
        .build_named();
    assert_eq!(expected, actual);
}
