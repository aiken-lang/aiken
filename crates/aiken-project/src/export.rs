use aiken_lang::{ast::TypedFunction, gen_uplc::CodeGenerator};
use miette::NamedSource;
use uplc::ast::{DeBruijn, Program};

use crate::{
    blueprint::{
        self,
        definitions::Definitions,
        parameter::Parameter,
        schema::{Annotated, Schema},
    },
    module::{CheckedModule, CheckedModules},
};

#[derive(Debug, PartialEq, Clone, serde::Serialize, serde::Deserialize)]
pub struct Export {
    pub name: String,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub doc: Option<String>,

    #[serde(skip_serializing_if = "Vec::is_empty")]
    #[serde(default)]
    pub parameters: Vec<Parameter>,

    #[serde(flatten)]
    pub program: Program<DeBruijn>,

    #[serde(skip_serializing_if = "Definitions::is_empty")]
    #[serde(default)]
    pub definitions: Definitions<Annotated<Schema>>,
}

impl Export {
    pub fn from_function(
        func: &TypedFunction,
        module: &CheckedModule,
        generator: &mut CodeGenerator,
        modules: &CheckedModules,
    ) -> Result<Export, blueprint::Error> {
        let mut definitions = Definitions::new();

        let parameters = func
            .arguments
            .iter()
            .map(|param| {
                Annotated::from_type(
                    modules.into(),
                    blueprint::validator::tipo_or_annotation(module, param),
                    &mut definitions,
                )
                .map(|schema| Parameter {
                    title: Some(param.arg_name.get_label()),
                    schema,
                })
                .map_err(|error| blueprint::Error::Schema {
                    error,
                    location: param.location,
                    source_code: NamedSource::new(
                        module.input_path.display().to_string(),
                        module.code.clone(),
                    ),
                })
            })
            .collect::<Result<_, _>>()?;

        let program = generator
            .generate_raw(&func.body, &func.arguments, &module.name)
            .to_debruijn()
            .unwrap();

        Ok(Export {
            name: format!("{}.{}", &module.name, &func.name),
            doc: func.doc.clone(),
            parameters,
            program,
            definitions,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::{CheckedModules, Export};
    use crate::tests::TestProject;
    use aiken_lang::{
        self,
        ast::{TraceLevel, Tracing},
    };

    macro_rules! assert_export {
        ($code:expr) => {
            let mut project = TestProject::new();

            let modules = CheckedModules::singleton(project.check(project.parse(indoc::indoc! { $code })));

            let mut generator = project.new_generator(
                Tracing::All(TraceLevel::Verbose),
            );

            let (module, func) = modules
                .functions()
                .next()
                .expect("source code did no yield any exports");

            let export = Export::from_function(func, module, &mut generator, &modules);

            match export {
                Err(e) => insta::with_settings!({
                    description => concat!("Code:\n\n", indoc::indoc! { $code }),
                    omit_expression => true
                }, {
                    insta::assert_debug_snapshot!(e);
                }),

                Ok(validator) => insta::with_settings!({
                    description => concat!("Code:\n\n", indoc::indoc! { $code }),
                    omit_expression => true
                }, {
                    insta::assert_json_snapshot!(validator);
                }),
            };
        };
    }

    #[test]
    fn basic_export() {
        assert_export!(
            r#"
            pub fn add(a: Int, b: Int) -> Int {
                a + b
            }
            "#
        );
    }

    #[test]
    fn illegal_opaque_type() {
        assert_export!(
            r#"
            pub opaque type Thing {
              a: Int
            }

            pub fn add(a: Thing, b: Int) -> Int {
                a.a + b
            }
            "#
        );
    }

    #[test]
    fn recursive_types() {
        assert_export!(
            r#"
            pub type Foo<a> {
              Empty
              Bar(a, Foo<a>)
            }

            pub fn add(a: Foo<Int>, b: Foo<Int>) -> Int {
              when (a, b) is {
                (Empty, Empty) -> 0
                (Bar(x, y), Bar(c, d)) -> x + c + add(y, d)
                (Empty, Bar(c, d)) -> c + add(Empty, d)
                (Bar(x, y), Empty) -> x + add(y, Empty)
              }
            }
            "#
        );
    }

    #[test]
    fn cannot_export_generics() {
        assert_export!(
            r#"
            pub fn add(_a: a, _b: b) -> Bool {
                True
            }
            "#
        );
    }
}
