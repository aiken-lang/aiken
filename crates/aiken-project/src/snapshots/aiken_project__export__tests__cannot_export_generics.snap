---
source: crates/aiken-project/src/export.rs
description: "Code:\n\npub fn add(_a: a, _b: b) -> Bool {\n    True\n}\n"
---
Schema {
    error: Error {
        context: FreeTypeVariable,
        breadcrumbs: [
            Var {
                tipo: RefCell {
                    value: Generic {
                        id: 33,
                    },
                },
                alias: None,
            },
        ],
    },
    location: 11..16,
    source_code: NamedSource {
        name: "",
        source: "<redacted>",
    ,
}
